apps/github/app_api.go

```diff
@@ -557,98 +562,369 @@ func (a AppAPI) processPullRequestEvent(ct context.Context, teamID uint64, evt e
 	if err != nil {
 		a.dataCollector.Logger.ErrorWithContext(ct, err)
 		return err
 	}
 
-	moveTaskToDeliveredRequest := &proto.MoveTaskToDeliveredRequest{
-		TaskId: pr.InternalTaskID,
-	}
-	_, rpcErr := a.teamyClientRegistry.TaskClient().MoveTaskToDelivered(ct, moveTaskToDeliveredRequest)
-	if rpcErr != nil {
-		internalErr := errs.FromGRPCErr(rpcErr)
-		a.dataCollector.Logger.ErrorWithContext(ct, internalErr)
-		return internalErr
+	for _, prTaskRelation := range prTaskRelations {
+		moveTaskToDeliveredRequest := &proto.MoveTaskToDeliveredRequest{
+			TaskId: prTaskRelation.InternalTaskID,
+		}
+		_, rpcErr := a.teamyClientRegistry.TaskClient().MoveTaskToDelivered(ct, moveTaskToDeliveredRequest)
+		if rpcErr != nil {
+			internalErr := errs.FromGRPCErr(rpcErr)
+			a.dataCollector.Logger.ErrorWithContext(ct, internalErr)
+			return internalErr
+		}
 	}
 
 	return nil
 }
 
-func (a AppAPI) createTaskForPullRequest(ct context.Context, teamID uint64, evt event, prEvt pullRequestEvent) *errs.Error {
-	prAuthorUserID, err := a.GetInternalUserID(ct, prEvt.PullRequest.User.NodeID)
+func (a AppAPI) closePullRequest(ct context.Context, teamID uint64, prEvt pullRequestEvent) *errs.Error {
+	prTaskRelations, err := a.githubPullRequestInternalTaskRelationDao.FindPullRequestInternalTaskRelationsByNodeID(ct, prEvt.PullRequest.NodeID)
 	if err != nil {
 		a.dataCollector.Logger.ErrorWithContext(ct, err)
 		return err
 	}
 
-	createTaskReq := &proto.CreateTaskRequest{
-		TeamId:      teamID,
-		Goal:        fmt.Sprintf("[%v][PR #%v] %v", evt.Repository.Name, prEvt.Number, prEvt.PullRequest.Title),
-		Context:     &prEvt.PullRequest.Body,
-		OwnerUserId: &prAuthorUserID,
+	if len(prTaskRelations) > 0 {
+		body := prEvt.PullRequest.Body
+		for _, prTaskRelation := range prTaskRelations {
+			err := a.removePullRequestTaskRelation(ct, prTaskRelation)
+			if err != nil {
+				a.dataCollector.Logger.ErrorWithContext(ct, err)
+				return err
+			}
+
+			body = strings.ReplaceAll(body, fmt.Sprintf("(task:%d)", prTaskRelation.InternalTaskID), "")
+		}
+
+		githubAppInstallation, err := a.githubAppInstallationDao.FindInstallationByTeamID(ct, teamID)
+		if err != nil {
+			a.dataCollector.Logger.ErrorWithContext(ct, err)
+			return err
+		}
+
+		installation := a.githubApp.GetInstallation(githubAppInstallation.ID)
+
+		_, err = a.githubGraphQLAPI.UpdatePullRequest(ct, installation, client.UpdatePullRequestInput{
+			PullRequestID: prEvt.PullRequest.NodeID,
+			Body:          &body,
+		})
+		if err != nil {
+			a.dataCollector.Logger.ErrorWithContext(ct, err)
+			return err
+		}
 	}
-	createTaskRes, rpcErr := a.teamyClientRegistry.TaskClient().CreateTask(ct, createTaskReq)
-	if rpcErr != nil {
-		internalErr := errs.FromGRPCErr(rpcErr)
-		a.dataCollector.Logger.ErrorWithContext(ct, internalErr)
-		return internalErr
+
+	return nil
+}
+
+func (a AppAPI) tryGetValidMentionedTasks(ct context.Context, body *string) (map[uint64]*proto.TaskMsg, *errs.Error) {
```

Comment:
```suggestion
func (a AppAPI) tryGetValidMentionedTasks(ct context.Context, body string) (map[uint64]entity.Task, *errs.Error) {
```

Let's convert `proto.TaskMsg` to internal `Task` model before consuming it in the rest of the logic.

Reply:
open this again, why we are doing this?